src/hooks/cgroups/pbs_cgroups.PY

```diff
@@ -0,0 +1,3657 @@
+                result.append(p)
+                continue
+            m = os.lstat(p).st_mode
+            if 'f' in kind and stat.S_ISREG(m):
+                result.append(p)
+            elif 'l' in kind and stat.S_ISLNK(m):
+                result.append(p)
+            elif 'c' in kind and stat.S_ISCHR(m):
+                result.append(p)
+            elif 'b' in kind and stat.S_ISBLK(m):
+                result.append(p)
+            elif 'p' in kind and stat.S_ISFIFO(m):
+                result.append(p)
+            elif 's' in kind and stat.S_ISSOCK(m):
+                result.append(p)
+        if name is None:
+            filtered = dirs
+        else:
+            filtered = fnmatch.filter(dirs, name)
+        for d in filtered:
+            p = os.path.join(root, d)
+            if kind is None:
+                result.append(p)
+                continue
+            m = os.lstat(p).st_mode
+            if 'd' in kind and stat.S_ISDIR(m):
+                result.append(p)
+            elif 'l' in kind and stat.S_ISLNK(m):
+                result.append(p)
+    return result
+
+
+# Return a properly cast zero value
+def initialize_resource(resc):
+    if isinstance(resc, pbs.pbs_int):
+        return pbs.pbs_int(0)
+    if isinstance(resc, pbs.pbs_float):
+        return pbs.pbs_float(0)
+    if isinstance(resc, pbs.size):
+        return pbs.size('0')
+    if isinstance(resc, int):
+        return 0
+    if isinstance(resc, float):
+        return 0.0
+    if isinstance(resc, list):
+        return []
+    if isinstance(resc, dict):
+        return {}
+    if isinstance(resc, tuple):
+        return ()
+    if isinstance(resc, str):
+        return ''
+    return resc
+
+
+# Helper function to ignore suspended jobs when removing
+# "already used" resources
+def job_to_be_ignored(jobid):
+    # Get job substate from small utility that calls printjob
+    pbs_exec = ''
+    if pbs_version[0] >= 13:
+        # This will fail for versions of PBS prior to 13.0
+        pbs_conf = pbs.get_pbs_conf()
+        pbs_exec = pbs_conf['PBS_EXEC']
+    else:
+        if 'PBS_EXEC' in self.cfg:
+            pbs_exec = self.cfg['PBS_EXEC']
+        else:
+            pbs.logmsg(pbs.EVENT_DEBUG2,
+                       "PBS_EXEC needs to be defined in the config file")
+            pbs.logmsg(pbs.EVENT_DEBUG2,
+                       "Exiting the cgroups hook")
+            pbs.event().accept()
+    printjob_cmd = os.path.join(pbs_exec, 'bin', 'printjob')
+    cmd = [printjob_cmd, jobid]
+    substate = None
+    try:
+        pbs.logmsg(pbs.EVENT_DEBUG3, "cmd: %s" % cmd)
+        # Collect the job substate information
+        process = subprocess.Popen(cmd, shell=False,
+                                   stdout=subprocess.PIPE,
+                                   stderr=subprocess.PIPE)
+        out, err = process.communicate()
+        # Find the job substate
+        substate_re = re.compile(r"substate:\s+(?P<jobstate>\S+)\s+")
```

Comment:
Not sure why it was done that way. Using group 'jobstate' prevents us from having to split and strip later.